### PR TITLE
storage: New circular buffer on non volatile memory

### PR DESCRIPTION
**Pull request by Laczen**
A new circular buffer mechanism (nvcb) is proposed for storing data on non volatile memory. The solution stores data as cbor binary string items. The non volatile memory is divided into configurable block sizes. Data is written sequentially to a block and when a block is full a new block can be taken into use. When the non volatile memory reserved for the circular buffer is full the first block is reused to store data.

More information on how to set it up can be found in `nvcb.h`.

In comparison with fcb the proposal provides the following:
a. Support for non flash backends, e.g. also eeproms can be used,
b. Content can easily be generated (writing cbor entries),
c. Support backends with non constant block size,

Backend configuration for flash and eeprom will follow shortly

This backend will be the base of a new key-value store that allows direct storage and retrieval of key-value items and maintains their validity when new items are added.

This new key-value store will then be added as a backend for settings.

### CODE REVIEW

File: subsys/storage/nvcb/nvcb.c [hunk 1]
```diff
@@ -0,0 +1,882 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/storage/nvcb.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(fs_nvcb, CONFIG_NVCB_LOG_LEVEL);
+
+#define NVCB_ENTRYHDRS        0b01011000 /* cbor byte string, 1 byte length */
```
Comment (rghaddab) [inline subsys/storage/nvcb/nvcb.c:14]: Does this mean that the entry length is always 24 bytes for this type of entry
According to https://datatracker.ietf.org/doc/html/rfc8949 0b010_xxxxx represents the Major type 2, where xxxxx is the length if it is <= 24

File: subsys/storage/nvcb/nvcb.c [hunk 2]
```diff
@@ -0,0 +1,882 @@
...
+		hdr[0] = NVCB_ENTRYHDRL;
+		hdr[1] = dlen >> 8;
+		hdr[2] = dlen;
+		hdr[3] = pass;
+		rv = 4U;
+		goto end;
+	}
+
+	hdr[0] = NVCB_ENTRYHDRS;
+	hdr[1] = dlen;
```
Comment (rghaddab) [inline subsys/storage/nvcb/nvcb.c:281]: for cbor entries less than 24 bytes, shouldn't be the length in the header ?

**Review by rghaddab — COMMENTED** [review]
Before I continue the review I just want to verify one thing.
This is not a cbor entries compatible like it is defined here
https://datatracker.ietf.org/doc/html/rfc8949
Am I wrong or you just wanted to implement something that looks like cbor entries but more convenient for the circular buffer ?

**Comment by Laczen** [comment]
> Before I continue the review I just want to verify one thing. This is not a cbor entries compatible like it is defined here https://datatracker.ietf.org/doc/html/rfc8949 Am I wrong or you just wanted to implement something that looks like cbor entries but more convenient for the circular buffer ?

@rghaddab thank you for taking the time to look at this. I first wanted to propose this to the Architecture Meeting to avoid any reviewers doing a review of something that would not be accepted anyhow (it is another storage solution for small items and we alreafy have fcb and nvs).

The reason to use cbor entries is to allow easy off chip generation of a set of data that would be used as an upload to a region on flash or even directly included in source code. There are libraries for python, c, ... to generate cbor (these libraries do not exist for fcb or nvs). The datatype chosen is cbor major type 2. This is a simple datatype that can also he used as the base for nvcb.

File: subsys/storage/nvcb/nvcb.c (hunk 1 above)
Comment (Laczen) [reply subsys/storage/nvcb/nvcb.c:14]: For length < 24 there are multiple Major type 2 that should be accepted. For bytestrings with length < 24 the length can be added to the header (most efficient), but also a 2 (or even a 4) byte  length specifier are valid representations. To fully support the off chip generation (that would generate the most efficient data size) it is better to add specific handling of these smaller items. I will add this as when it is clear that the proposed nvcb will not be rejected.

**Pull request closed without merging**